### ISSUE TITLE
[dtest, tablets] test_create_lots_of_indexes_concurrently fails with tablets: Expected a row count of 1 in query "select * from base_X where c2 = Y", but got 0

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#17603]
See https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/dtest-byo/143/testReport/concurrent_schema_changes_test/TestConcurrentSchemaChanges/FullDtest___full_split002___test_create_lots_of_indexes_concurrently/
```
self = <concurrent_schema_changes_test.TestConcurrentSchemaChanges object at 0x7f536f5a5b10>
fixture_dtest_setup = <dtest_setup.DTestSetup object at 0x7f536c4059d0>

    def test_create_lots_of_indexes_concurrently(self, fixture_dtest_setup):
        """
        create indexes across multiple threads concurrently
        """
        cluster = self.cluster
        cluster.populate(2).start()
    
        node1, node2 = cluster.nodelist()
        session = self.cql_connection(node1)
        session.execute("create keyspace lots_o_indexes WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1};")
        session.execute("use lots_o_indexes")
        for n in range(5):
            session.execute(f"create table base_{n} (id uuid primary key, c1 int, c2 int)")
            for ins in range(1000):
                session.execute(f"insert into base_{n} (id, c1, c2) values (uuid(), {ins}, {ins})")
        wait(5)
    
        fixture_dtest_setup.ignore_log_patterns += [
            r"view - (\(rate limiting dropped [0-9]+ similar messages\) )?Error applying view update to .*: exceptions::mutation_write_failure_exception",
            r"view - (\(rate limiting dropped [0-9]+ similar messages\) )?Failed to apply mutation from .*: data_dictionary::no_such_column_family",
        ]
    
        logger.debug("creating indexes")
        cmds = []
        for n in range(5):
            cmds.append((f"create index ix_base_{n}_c1 on base_{n} (c1)", ()))
            cmds.append((f"create index ix_base_{n}_c2 on base_{n} (c2)", ()))
    
        results = execute_concurrent(session, cmds, raise_on_first_error=True)
    
        for success, result in results:
            assert success, f"didn't get success on table create: {result}"
    
        wait(5)
    
        logger.debug("validating schema and index list")
        session.cluster.control_connection.wait_for_schema_agreement()
        session.cluster.refresh_schema_metadata()
        index_meta = session.cluster.metadata.keyspaces["lots_o_indexes"].indexes
        self.validate_schema_consistent(node1)
        self.validate_schema_consistent(node2, timeout=0)
        assert 10 == len(index_meta), f"expect 10 , got len(index_meta)={len(index_meta)}"
        for n in range(5):
            assert f"ix_base_{n}_c1" in index_meta, f"ix_base_{n}_c1 not found in {index_meta}"
            assert f"ix_base_{n}_c2" in index_meta, f"ix_base_{n}_c2 not found in in {index_meta}"
    
        logger.debug("waiting for indexes to fill in")
        wait(45)
        logger.debug("querying all values by secondary index")
        for n in range(5):
            for ins in range(1000):
                assert_row_count_in_select(session, f"select * from base_{n} where c1 = {ins}", 1)
>               assert_row_count_in_select(session, f"select * from base_{n} where c2 = {ins}", 1)

concurrent_schema_changes_test.py:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tools/retrying.py:40: in inner
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <cassandra.cluster.Session object at 0x7f536465fcd0>
query = 'select * from base_2 where c2 = 12', num_rows_expected = 1
consistency_level = 1, num_attempts = 1, timeout = None

    @retrying(num_attempts=1, sleep_time=10)
    def assert_row_count_in_select(  # noqa: PLR0913
        session,
        query,
        num_rows_expected,
        consistency_level=ConsistencyLevel.ONE,
        num_attempts=1,
        timeout=None,
    ):
        """
        Function to validate the row count are returned by select
        :param num_attempts: defines how many time try to assert data in case failure. Used in retrying decorator
        """
        from tools.data import get_list_res  # to avoid cyclic dependency
    
        count = len(get_list_res(session, query, consistency_level, timeout=timeout))
>       assert count == num_rows_expected, f'Expected a row count of {num_rows_expected} in query "{query}", but got {count}'
E       AssertionError: Expected a row count of 1 in query "select * from base_2 where c2 = 12", but got 0

tools/assertions.py:517: AssertionError
```

**Comment by bhalevy** [comment]
@nyh can you please look into this issue?
I'll disable it for tablets in dtest for now, mentioning this issue.

**Comment by nyh** [comment]
The test has problems, but I think it may reproduce a real bug - maybe https://github.com/scylladb/scylladb/issues/17297 but maybe even something different:

The test creates five tables with 1000 rows in each. It then creates a secondary index on each of these five tables (these five creations happen in parallel, apparently this is the point (?) of this test. It then sleeps 50 seconds expects all the data to be searchable in the secondary indexes. The test fails when the value 12 is not present in the index of the second table.
 
On one hand, it's wrong to sleep for "50 seconds" assuming it should always be enough. In this sense, this test is flawed.
On the other hand, 50 seconds should be enough for building just 5,000 view rows, and the logs show that it is - for example in https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/143/artifact/logs-full.release.002/1708425661854_concurrent_schema_changes_test.py%3A%3ATestConcurrentSchemaChanges%3A%3Atest_create_lots_of_indexes_concurrently/node2.log
```
INFO  2024-02-20 10:40:03,275 [shard 0:strm] schema_tables - Creating lots_o_indexes.ix_base_2_c2_index id=67933741-cfdc-11ee-a10f-c99414289ceb version=67933742-cfdc-11ee-a10f-c99414289ceb
INFO  2024-02-20 10:40:03,303 [shard 0:strm] view - Building view lots_o_indexes.ix_base_2_c2_index, starting at token -6991559177879901395
INFO  2024-02-20 10:40:03,861 [shard 0:strm] view - Finished building view lots_o_indexes.ix_base_2_c2_index
```

The view building of this index of c2 over base_2 (the index needed in the query) finished about half a second after the index was created, 50 seconds is more than enough.

This is likely to not be the same as https://github.com/scylladb/scylladb/issues/17297 and be a different bug because the logs of both nodes (see
https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/143/artifact/logs-full.release.002/1708425661854_concurrent_schema_changes_test.py%3A%3ATestConcurrentSchemaChanges%3A%3Atest_create_lots_of_indexes_concurrently/ ) seem to suggest that the building of this index finished successfully. So it seems in first glance that the view build didn't hang, but missed some of the data.

**Comment by mykaul** [comment]
@temichus - do you know if this is still relevant?

**Comment by temichus** [comment]
> @temichus - do you know if this is still relevant?

I did not know about this issue.

but as I see test still disabled for tablets due to this issue 
```
@pytest.mark.required_features("!tablets")  # Due to https://github.com/scylladb/scylladb/issues/17603
```

and still has  sleep for "50 seconds" that Nadav mentioned

**Comment by mlitvk** [comment]
I wasn't able to reproduce this in thousands of normal test runs, but I was able to reproduce it by forcing tablet movement while the index is building, which may be what happened in the failed run.
I think there is a known consistency issue of views with tablets, maybe https://github.com/scylladb/scylladb/issues/18826, and the vb coordinator addresses it.

**Comment by mlitvk** [comment]
more elaboration on why we generate view updates on the wrong replica:
I think the scenario is:
1. view builder initializes a reader of the local sstables
2. the tablet starts migrating to another node
3. the tablet stage is set to use_new => the migration source is not a replica anymore
4. view builder continues to generate view update based on the local sstable reader, but now we get base tokens for which we are not a replica, and it fails
5. I think the migration target doesn't generate view updates for these "bad" base tokens as well, because maybe it also has an outdated reader

I suppose we don't hold the token map while having this reader. I'm not sure if it's a problem for vnodes, if a node joins or leaves during view building?

For tablets, I think the issue is addressed by the view building coordinator. when a tablet migrates, it moves the view building task to the target replica.

**Comment by mlitvk** [comment]
I tried to check what happens with vnodes when the token metadata changes during view building
Suppose during view building a new node joins, so some of our tokens move to other nodes while we generate view updates for them.
I was able to reproduce the `Could not find ... in base_endpoints` error in this scenario, but I didn't get missing view rows. maybe there are generated during streaming?
so I'm not sure that it's a problem for vnodes

**Comment by piodul** [comment]
@wmitros / @Jadw1 please comment on the two most recent comments, you should be more up-to-date with regards to the legacy view builder and how it handles readers / token metadata

**Comment by mlitvk** [comment]
I looked more into view update generation with streaming:
* with vnodes, when a node joins, I think there is a repair which causes us to generate view updates for the problematic tokens
* with tablets, it uses file streaming, and I think it doesn't use the staging sstables to generate view updates?
* with tablets and file streaming disabled, I see it does register the sstable as staging for view update generation, but we still get missing view rows

**Comment by mykaul** [comment]
Thanks @mlitvk for the excellent analysis. I suspect we should split this issue to tablets and vnodes?

**Comment by mlitvk** [comment]
so I'm not sure if there's an issue to address
with vnodes I didn't encounter a consistency issue
with tablets the implementation is going to be changed with the view building coordinator. @Jadw1 can you please see if it addresses the issues above?

**Comment by dani-tweig** [comment]
> The issue is already addressed in [#23760](https://github.com/scylladb/scylladb/pull/23760). When a tablet is migrated, for any unfinished view building task, a new task will be created on the new replica and the coordinator will start to build the tablet only when the migration is fully finished (the tablet has no transition info).

Can we close this one then?

**Comment by mykaul** [comment]
> > The issue is already addressed in [#23760](https://github.com/scylladb/scylladb/pull/23760). When a tablet is migrated, for any unfinished view building task, a new task will be created on the new replica and the coordinator will start to build the tablet only when the migration is fully finished (the tablet has no transition info).
> 
> Can we close this one then?

Unlikely, based on the discussion @ https://github.com/scylladb/scylladb/issues/15046

**Comment by Jadw1** [comment]
I can add "Fixes ..." tag in the https://github.com/scylladb/scylladb/pull/23760, since this PR will be fixing it
